### PR TITLE
(2727) Reject ODA fields in non-ODA activity upload

### PR DESCRIPTION
**Pull request by yndajas**
If a user uploads a completed ODA bulk upload template using the non-ODA form, we don't want to create a non-ODA activity with the additional fields, but instead reject the upload since it doesn't match out expectations for ODA typing. This safeguards against uploading activities as the wrong type by mistake. This only seems to be possible in one direction, since the ODA templates have all the non-ODA fields and more, while the non-ODA templates don't include various required fields for ODA

This compares the ODA and non-ODA attributes and adds an error if any unexpected attributes appear in the non-ODA upload

### After

image

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

File: app/services/activity/import.rb [hunk 1]
```diff
@@ -219,7 +256,17 @@ def create
...
+        attributes = @converter.to_h
+
+        if @is_oda == false
+          incorrect_oda_type_errors = Activity::Import.incorrect_oda_type_errors(@activity, attributes)
+          @errors.merge!(incorrect_oda_type_errors)
+        end
+
+        return unless @errors.blank?
+
+        @activity.assign_attributes(attributes)
```
Comment (yndajas) [inline app/services/activity/import.rb:269]: If comparing this with the changes in `ActivityUpdater` above, you can see this isn't very DRY (except the `Activity::Import.incorrect_oda_type_errors` abstraction), but neither is some of the other code in these methods, so I've added a tech debt card

**Comment by yndajas** [comment]
[Ensure you can't upload an ODA activity via the non-ODA bulk upload](https://trello.com/c/E8jAGO7O/2727-ensure-you-cant-upload-an-oda-activity-via-the-non-oda-bulk-upload)

File: app/services/activity/import.rb [hunk 2]
```diff
@@ -30,6 +30,34 @@ def is_oda_by_type(type:)
           non_ispf: nil
         }[type]
       end
+
+      def unexpected_non_oda_attribute_errors(activity, converted_attributes)
+        level = {
+          "programme" => :level_b,
+          "project" => :level_c_d,
+          "third_party_project" => :level_c_d
+        }[activity.level]
```
Comment (yndajas) [inline app/services/activity/import.rb:39]: Maybe we should just pass in the level like we do with `filtered_csv_column_headings`

File: spec/services/activity/import_spec.rb [hunk 3]
```diff
@@ -333,6 +333,107 @@
...
+    }
+
+    context "Level B" do
+      let(:activity) { create(:programme_activity, :ispf_funded, is_oda: true) }
+      let(:error_translation) { t("importer.errors.activity.oda_attribute_in_non_oda_activity") }
+
+      it "returns a hash with errors where ODA fields have been included in a non-ODA upload" do
+        expect(Activity::Import.unexpected_non_oda_attribute_errors(activity, converted_level_b_oda_attributes)).to eq(
+          {
+            aid_type: ["D01", error_translation],
```
Comment (yndajas) [inline spec/services/activity/import_spec.rb:372]: `converted_level_b_oda_attributes.aid_type` be cleaner than `"D01"`?

**Review by CristinaRO — APPROVED** [review]
One non-blocking comment/suggestion.

**Review by Gweaton — APPROVED** [review]
Looks good - I think it'd be good to, as Cristina says, move that calculation of the unexpected fields somewhere else. I found it a little hard to parse at first read too, so it might improve readability and simplify the method too.

File: app/services/activity/import.rb [hunk 4]
```diff
@@ -30,6 +30,34 @@ def is_oda_by_type(type:)
...
+
+        non_oda_attributes = []
+        oda_attributes = []
+
+        ACTIVITY_CSV_COLUMNS.each do |attribute, column|
+          non_oda_attributes << attribute if column.dig(:inclusion, level, :ispf_non_oda)
+          oda_attributes << attribute if column.dig(:inclusion, level, :ispf_oda)
+        end
+
+        unexpected_attributes = oda_attributes - non_oda_attributes
```
[the remark this answers is not in the document]
Comment (yndajas) [reply app/services/activity/import.rb]: Per Slack, have added a commit for review showing what this might look like

File: app/services/activity/import.rb [hunk 5]
```diff
@@ -131,7 +146,21 @@ def initialize(row:, uploader:, partner_organisation:, report:, is_oda:)
       def update
         return unless @activity && @errors.empty?
 
-        @activity.assign_attributes(@converter.to_h)
+        attributes = @converter.to_h
+
+        if @is_oda == false
+          level = @activity.programme? ? :level_b : :level_c_d
```
Comment (Gweaton) [inline app/services/activity/import.rb:152]: Sorry - I didn't realise we'd need to convert the level into a symbol here to then be checked in the method - I think it might be cleaner to pass in the `activity` itself here and call `.programme?` inside `unexpected_non_oda_attribute_errors` here.

File: config/initializers/activity_csv_columns.rb [hunk 6]
```diff
@@ -2,3 +2,18 @@
...
+level_c_d_oda_attributes = []
+
+ACTIVITY_CSV_COLUMNS.each do |attribute, column|
+  level_b_non_oda_attributes << attribute if column.dig(:inclusion, :level_b, :ispf_non_oda)
+  level_b_oda_attributes << attribute if column.dig(:inclusion, :level_b, :ispf_oda)
+  level_c_d_non_oda_attributes << attribute if column.dig(:inclusion, :level_c_d, :ispf_non_oda)
+  level_c_d_oda_attributes << attribute if column.dig(:inclusion, :level_c_d, :ispf_oda)
+end
+
+INVALID_LEVEL_B_ISPF_NON_ODA_ATTRIBUTES = level_b_oda_attributes - level_b_non_oda_attributes
```
Comment (Gweaton) [inline config/initializers/activity_csv_columns.rb:18]: Shall we `.freeze` these to be extra safe?

**Review by Gweaton — APPROVED** [review]
Thanks for updating this - a couple of small changes but good to merge afterwards (after squashing!)

File: app/services/activity/import.rb (hunk 5 above)
Comment (yndajas) [reply app/services/activity/import.rb:152]: I was conflicted about whether to pass it in as the symbol or `@activity.level` - I went with the symbol to keep it consistent with `filtered_csv_column_headings`, but it might be cleaner to pass the `@activity.level` and do the conversion inside (like before, but without the whole activity), or yeah, pass the whole activity and use the `programme?` method. Any preference?

File: config/initializers/activity_csv_columns.rb (hunk 6 above)
Comment (yndajas) [reply config/initializers/activity_csv_columns.rb:18]: Sure!

Comment (yndajas) [reply config/initializers/activity_csv_columns.rb]: Does this approach more generally seem sensible to you? If so, I'll squash the commits

Comment (Gweaton) [reply config/initializers/activity_csv_columns.rb]: I think so - I sort of wish there was a way that didn't involve looping all over and subtracting (like if we could just do something like `columns.where(level_b: true, ispf_oda: true, ispf_non_oda: false)`, but I can't think of a neat way of doing that with the way our hash is structured right now. Let's go with this for now.

File: app/services/activity/import.rb (hunk 5 above)
Comment (Gweaton) [reply app/services/activity/import.rb:152]: I think passing in and use the `programme?` would be my preference, but up to you - apologies again for flip-flopping on this!

File: config/initializers/activity_csv_columns.rb (hunk 6 above)
Comment (yndajas) [reply config/initializers/activity_csv_columns.rb]: Yeah, `filtered_csv_column_headings` is essentially that kind of interface but for headers, not attributes. I suppose we could make an almost identical `filtered_csv_column_attributes` (or something more accurate) but the only use case right now is working out the difference
